### ISSUE TITLE
are the roles of x_i and x_j swapped in MessagePassing object method message?

### ISSUE THREAD

**Issue by atmamta2** [rusty1s/pytorch_geometric#699]
# Hi,
# it seems to me that the roles of x_i (source nodes) and x_j (destination nodes) are mixed up in  nn.MessagePassing objects. Here is a code segment that demonstrates the issue:

class SrcDstMean(MessagePassing):
    """Calculates mean of one feature along the edges."""
    def __init__(self):
        super().__init__(aggr="mean", flow="source_to_target") 

    def forward(self, feat, edge_index):
        return self.propagate(edge_index, x=feat)

    def message(self, x_i, x_j, edge_index, size):
        return  torch.stack([x_i, x_j], dim=1)
        
    def update(self, aggr_out):
        return aggr_out
    
sdmean = SrcDstMean ()
# graph nodes form a single chain: 0 -> 1 -> 2 -> 3 -> 4 -> 5, where arrows stand for edges.
index_arr = torch.LongTensor ([[0,1,2,3,4], [1,2,3,4,5]])
res = sdmean (torch.Tensor ([z*z+100. for z in range (6)]), index_arr)
res

# this yields the following:
tensor([[  0.,   0.],
             [101., 100.],
             [104., 101.],
             [109., 104.],
             [116., 109.],
             [125., 116.]])

# The first column is the vector made of the x_i-s, the second is the vector of x_j-s. Clearly, x_i are the  destination indicies, whereas they should be the source, according to the docs... Or am I missing  something here?

# thx!

**Comment by WMF1997** [comment]
to show more clearly, I retyped the code: (and i modified some...)

```python
import torch
import torch_geometric
from torch_geometric.nn import MessagePassing
class SrcDstMean(MessagePassing):
    """Calculates mean of one feature along the edges."""
    def init(self):
        super().init(aggr="mean", flow="source_to_target")
    def forward(self, x, edge_index):
        return self.propagate(edge_index, x=x)
    def message(self, x_i, x_j, edge_index, size):
        return  torch.stack([x_i, x_j**2], dim=1)
    def update(self, aggr_out):
        return aggr_out + 2333
if __name__ == "__main__":
    sdmean = SrcDstMean()
    index_arr = torch.LongTensor([[0,1,2,3,4], [1,2,3,4,5],])
    x = torch.Tensor([z**2+100. for z in range(6)])
    res = sdmean(x=x, edge_index=index_arr)
```

```
In [8]: res
Out[8]: 
tensor([[ 2333.,  2333.],
        [ 2434., 12333.],
        [ 2437., 12534.],
        [ 2442., 13149.],
        [ 2449., 14214.],
        [ 2458., 15789.]])
```

**Comment by rusty1s** [comment]
We generally refer to `x_i` as the nodes which aggregate information, and `x_j` which send information along the edges.

The default message passing flow is `source_to_target`, so `x_i` refers to the targets and `x_j` refers to the source nodes. If you want to change this behavior, you can change the message passing flow via `flow="target_to_source".

**Comment by atmamta2** [comment]
thanks for the quick reply. 
Ok, that is good to know. It would be great if this could be explained more clearly in the documentation as well. This is what the docs say now:
*************************************************************
 message(x_j)[source]
Constructs messages in analogy to ϕΘ
for each edge in (i,j)∈E. Can take any argument which was initially passed to propagate(). In addition, features can be lifted to the source node i and target node j by appending _i or _j to the variable name, .e.g. x_i and x_j.
************************************************************
Now (i,j) is in E, i.e. (i,j) in the edge set means (at least in graph theory) that there is a directed edge going from node i to node j. So this is very, very misleading...

**Comment by rusty1s** [comment]
I tried to make this more clear in the documentation. For `"source_to_target"`, edges are now encoded as `(j,i)`.

**Comment by czha5168** [comment]
> I tried to make this more clear in the documentation. For `"source_to_target"`, edges are now encoded as `(j,i)`.

Dear Matthias,
Here might be a typo on https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html?
=> Inside the message() function, we use self.mlp to transform both the source node features x_i and the relative target node features x_j - x_i for each edge.
=> Should it be the target node feature x_i and the source node feature x_j?
Thanks!
